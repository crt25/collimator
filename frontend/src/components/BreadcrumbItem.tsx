import Link from "next/link";

const BreadcrumbItem = ({
  href,
  children,
}: {
  href?: string;
  children: React.ReactNode;
}) => (
  <li className="breadcrumb-item">
    {href ? <Link href={href}>{children}</Link> : children}
  </li>
);

export default BreadcrumbItem;
