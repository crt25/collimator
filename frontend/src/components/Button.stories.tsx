import { fn } from "@storybook/test";
import Button from "./Button";

export default {
  component: Button,
  title: "Button",
};

export const Default = {
  args: {
    children: "Some string content",
  },
};

export const RichContent = {
  args: {
    children: (
      <>
        <u>hello</u> <strong>world</strong>
      </>
    ),
  },
};

export const WithSucceedingPromise = {
  args: {
    children: "easy task taking 5s",
    onClick: fn(
      () =>
        new Promise<void>((resolve) => {
          setTimeout(resolve, 5 * 1000);
        }),
    ),
  },
};

export const WithFailingPromise = {
  args: {
    children: "impossible task taking 2s",
    onClick: fn(
      () =>
        new Promise((_, reject) => {
          setTimeout(reject, 2 * 1000);
        }),
    ),
  },
};

export const WithNeverReturningPromise = {
  args: {
    children: "what is the answer to life the universe and everything?",
    onClick: fn(() => new Promise(() => {})),
  },
};
