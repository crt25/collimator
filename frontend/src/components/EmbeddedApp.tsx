import {
  AppIFrameMessage,
  AppIFrameRequest,
  AppIFrameResponse,
} from "@/types/app-iframe-message";
import styled from "@emotion/styled";
import {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from "react";

const StyledIFrame = styled.iframe`
  width: 100%;
  border: none;

  flex-grow: 1;
`;

const MAX_COUNTER = 1000000;
let counter = 0;

const postMessageToIFrame = (
  iFrame: HTMLIFrameElement,
  message: AppIFrameMessage,
) => {
  if (!iFrame.contentWindow) {
    console.error(
      "Cannot post message to iframe without content window:",
      iFrame,
    );
    return;
  }

  // get target origin from the iframe's src attribute
  const targetOrigin = new URL(iFrame.src).origin;

  iFrame.contentWindow.postMessage(message, {
    targetOrigin,
  });
};

export interface EmbeddedAppRef {
  sendRequest<ProcedureName extends AppIFrameRequest["procedure"]>(
    request: Omit<AppIFrameRequest, "id" | "type"> & {
      procedure: ProcedureName;
    },
  ): Promise<AppIFrameResponse & { procedure: ProcedureName }>;
}

interface Props {
  src: string;
}

const EmbeddedApp = forwardRef<EmbeddedAppRef, Props>(function EmbeddedApp(
  { src },
  ref,
) {
  const [iFrame, setIFrame] = useState<HTMLIFrameElement | null>(null);
  const isIFrameLoaded = useRef<boolean>(false);

  const pendingRequests = useRef<{
    [key: number]: (response: AppIFrameResponse) => void;
  }>({});

  const sendRequest = useCallback(
    <ProcedureName extends AppIFrameMessage["procedure"]>(
      request: Omit<AppIFrameRequest, "id" | "type"> & {
        procedure: ProcedureName;
      },
    ): Promise<AppIFrameResponse & { procedure: ProcedureName }> => {
      if (iFrame && isIFrameLoaded.current) {
        return new Promise((resolve) => {
          // store the resolve function in the pendingRequests object
          pendingRequests.current[counter] = (response: AppIFrameResponse) => {
            if (response.procedure !== request.procedure) {
              console.error("Invalid response procedure", response, request);
              return;
            }

            resolve(
              response as AppIFrameResponse & { procedure: ProcedureName },
            );
          };

          // send the message to the iFrame
          postMessageToIFrame(
            iFrame,
            // add a unique id to the message
            {
              id: counter,
              type: "request",
              ...request,
            } as AppIFrameMessage,
          );

          // increment the counter
          counter = (counter + 1) % MAX_COUNTER;
        });
      }

      return Promise.reject("No iFrame available");
    },
    [iFrame],
  );

  useImperativeHandle(
    ref,
    () => ({
      sendRequest,
    }),
    [sendRequest],
  );

  // a callback function that listens for messages from the iFrame
  const handleIFrameMessage = useCallback(
    (event: MessageEvent) => {
      if (event.source !== iFrame?.contentWindow) {
        return;
      }

      const message = event.data as AppIFrameMessage;

      if (message.type === "request") {
        console.error("Requests are not supported");
        return;
      }

      // get the resolve function from the pendingRequests object
      const resolve = pendingRequests.current[message.id];
      if (!resolve) {
        console.error("No resolve function found for message", message);
      }

      // call the resolve function with the message
      resolve(message);
      // remove the resolve function from the pendingRequests object
      delete pendingRequests.current[message.id];
    },
    [iFrame],
  );

  // add an event listener to listen for messages from the iFrame
  useEffect(() => {
    window.addEventListener("message", handleIFrameMessage);

    return () => {
      window.removeEventListener("message", handleIFrameMessage);
    };
  }, [handleIFrameMessage]);

  // after the iFrame has been rendered, send a message to the iFrame
  useEffect(() => {
    if (iFrame) {
      const callback = async () => {
        isIFrameLoaded.current = true;

        const response = await sendRequest({
          procedure: "getHeight",
        });

        iFrame.style.height = `${response.result}px`;
      };
      iFrame.addEventListener("load", callback);

      return () => {
        if (iFrame) {
          iFrame.removeEventListener("load", callback);
        }
      };
    }
  }, [iFrame, sendRequest]);

  // store the reference to the iFrame element in a state value
  // this is only called once when the iFrame is mounted
  // and once when it is unmounted due to the empty dependency array
  const getIFramRef = useCallback((node: HTMLIFrameElement | null) => {
    if (node !== null) {
      setIFrame(node);
    }

    // see https://legacy.reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <StyledIFrame
      src={src}
      ref={getIFramRef}
      allow="fullscreen; camera; microphone"
      sandbox="allow-scripts allow-same-origin allow-downloads"
      referrerPolicy="no-referrer-when-downgrade"
      loading="lazy"
    />
  );
});

export default EmbeddedApp;
