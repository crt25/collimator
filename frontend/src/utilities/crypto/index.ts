export { decodeBase64, encodeBase64Url } from "./base64";
export { default as EphemeralKey } from "./EphemeralKey";
export { default as KeyPair } from "./KeyPair";
export { PasswordDerivedKey } from "./PasswordDerivedKey";
export { default as SymmetricKey } from "./SymmetricKey";
export { default as StudentKeyPair } from "./StudentKeyPair";
export { default as TeacherKeyPair } from "./TeacherLongTermKeyPair";
