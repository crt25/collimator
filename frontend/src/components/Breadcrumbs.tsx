import styled from "@emotion/styled";
import BreadcrumbItem from "./BreadcrumbItem";
import { FormattedMessage } from "react-intl";

const StyledNav = styled.nav`
  margin-top: 1rem;
  padding: 0.5rem;
  border: var(--foreground-color) 1px solid;
  border-radius: var(--border-radius);
`;

const StyledOl = styled.ol`
  margin: 0;
`;

const Breadcrumbs = ({
  homeHref,
  children,
}: {
  homeHref?: string;
  children?: React.ReactNode;
}) => (
  <StyledNav aria-label="breadcrumb">
    <StyledOl className="breadcrumb">
      <BreadcrumbItem href={homeHref ?? "/"}>
        <FormattedMessage id="Breadcrumbs.home" defaultMessage="Home" />
      </BreadcrumbItem>
      {children}
    </StyledOl>
  </StyledNav>
);

export default Breadcrumbs;
