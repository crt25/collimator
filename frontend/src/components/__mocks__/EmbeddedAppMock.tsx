import { forwardRef, useImperativeHandle } from "react";
import { EmbeddedAppRef, Props } from "../EmbeddedApp";

export const createMockedEmbeddedApp = (
  sendRequest: EmbeddedAppRef["sendRequest"],
) =>
  forwardRef<EmbeddedAppRef, Props>(function MockEmbeddedApp({ src }, ref) {
    useImperativeHandle(
      ref,
      () => ({
        sendRequest,
      }),
      [],
    );

    return <iframe src={src} />;
  });
