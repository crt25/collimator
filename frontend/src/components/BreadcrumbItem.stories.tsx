import BreadcrumbItem from "./BreadcrumbItem";
import Breadcrumbs from "./Breadcrumbs";

type Args = Parameters<typeof BreadcrumbItem>[0];

export default {
  component: BreadcrumbItem,
  title: "BreadcrumbItem",
  render: (args: Args) => (
    <Breadcrumbs>
      <BreadcrumbItem {...args} />
    </Breadcrumbs>
  ),
};

export const Default = {
  args: {
    href: "/",
    children: "Some string content",
  } as Args,
};
