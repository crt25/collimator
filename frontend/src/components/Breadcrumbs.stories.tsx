import BreadcrumbItem from "./BreadcrumbItem";
import Breadcrumbs from "./Breadcrumbs";

type Args = Parameters<typeof Breadcrumbs>[0];

export default {
  component: Breadcrumbs,
  title: "Breadcrumbs",
  render: (args: Args) => (
    <Breadcrumbs {...args}>
      <BreadcrumbItem href="#">Users</BreadcrumbItem>
      <BreadcrumbItem href="#">John</BreadcrumbItem>
      <BreadcrumbItem href="#">Details</BreadcrumbItem>
      <BreadcrumbItem href="#">Edit</BreadcrumbItem>
    </Breadcrumbs>
  ),
};

export const Default = {
  args: {
    homeHref: "/",
  } as Omit<Args, "children">,
};
