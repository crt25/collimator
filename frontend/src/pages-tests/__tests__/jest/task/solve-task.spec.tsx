import { createMockedEmbeddedApp } from "@/components/__mocks__/EmbeddedAppMock";
import { EmbeddedAppRef } from "@/components/EmbeddedApp";
import SolveTaskPage from "@/pages/session/[sessionId]/task/[taskId]/solve";
import { render } from "@testing-library/react";
import { IntlProvider } from "react-intl";
import English from "../../../../../content/compiled-locales/en.json";
import {
  AuthenticationContext,
  AuthenticationContextType,
} from "@/contexts/AuthenticationContext";
import { getFullyAuthenticatedStudentContext } from "@/contexts/__tests__/mock-contexts";
import { subtle } from "crypto";

const renderComponent = (context: AuthenticationContextType) => (
  <IntlProvider locale="en" messages={English}>
    <AuthenticationContext.Provider value={context}>
      <SolveTaskPage />
    </AuthenticationContext.Provider>
  </IntlProvider>
);

jest.mock("@/components/EmbeddedApp", () => ({
  __esModule: true,
  _getMockComponent: jest.fn(() => {
    throw new Error("Not implemented");
  }),
  get default() {
    return this._getMockComponent();
  },
}));

const EmbeddedAppMock = jest.requireMock("@/components/EmbeddedApp")
  ._getMockComponent as jest.Mock;

const EmbeddedApp = jest.requireActual("@/components/EmbeddedApp").default;

jest.mock("@/utilities/download", () => ({
  downloadBlob: jest.fn(),
}));

const downloadBlob = jest.requireMock("@/utilities/download")
  .downloadBlob as jest.Mock;

jest.mock("next/router", () => ({
  useRouter: jest.fn(() => ({
    query: { sessionId: 10, taskId: 5 },
  })),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

const crypto = subtle as SubtleCrypto;

describe("Solve Task Page", () => {
  it("renders page unchanged", async () => {
    EmbeddedAppMock.mockImplementation(() => EmbeddedApp);
    const context = await getFullyAuthenticatedStudentContext(crypto);

    const { container } = render(renderComponent(context));
    expect(container).toMatchSnapshot();
  });

  it("can obtain submissions from the app", async () => {
    const mockSendRequest = jest.fn<
      ReturnType<EmbeddedAppRef["sendRequest"]>,
      Parameters<EmbeddedAppRef["sendRequest"]>
    >(async () => ({
      id: 1,
      type: "response",
      procedure: "getSubmission",
      result: new Blob(["{}"], { type: "application/json" }),
    })) as EmbeddedAppRef["sendRequest"];

    EmbeddedAppMock.mockImplementation(() =>
      createMockedEmbeddedApp(mockSendRequest),
    );

    const context = await getFullyAuthenticatedStudentContext(crypto);

    const { findByTestId } = render(renderComponent(context));

    const submitButton = await findByTestId("submit-solution-button");
    expect(submitButton).toBeTruthy();

    submitButton.click();

    expect(mockSendRequest).toHaveBeenCalledTimes(1);
    expect(mockSendRequest).toHaveBeenCalledWith({
      procedure: "getSubmission",
    });
  });

  describe("while the iframe is loading", () => {
    const NeverLoadingIFrame = () => <div />;

    it("clicking the submit button is a noop", async () => {
      EmbeddedAppMock.mockImplementation(() => NeverLoadingIFrame);

      const context = await getFullyAuthenticatedStudentContext(crypto);

      const { findByTestId } = render(renderComponent(context));

      const submitButton = await findByTestId("submit-solution-button");
      expect(submitButton).toBeTruthy();

      submitButton.click();

      expect(downloadBlob).toHaveBeenCalledTimes(0);
    });
  });
});
