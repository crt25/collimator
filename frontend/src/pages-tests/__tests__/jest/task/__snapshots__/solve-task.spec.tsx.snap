// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Solve Task Page renders page unchanged 1`] = `
<div>
  <div
    class="css-to24xl"
  >
    <header
      class="css-vi10jy"
    >
      <div
        class="css-o2j9ze container"
      >
        <div
          class="css-p7rs9i"
        >
          <img
            alt="Logo"
            src="https://placeholder.pics/svg/300x100"
          />
        </div>
        <menu
          class="css-1w18ink"
        >
          <li>
            <button
              class="css-1aktket"
              data-testid="toggle-session-menu-button"
            >
              <span>
                Show Session
              </span>
            </button>
          </li>
          <li>
            <button
              class="css-1aktket"
              data-testid="submit-solution-button"
            >
              Submit Solution
            </button>
          </li>
          <li>
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                class="dropdown-toggle btn btn-primary"
                id="react-aria-:r0:"
                type="button"
              >
                john@doe.com
              </button>
            </div>
          </li>
        </menu>
      </div>
    </header>
    <div
      class="css-qrtzlc"
    >
      <iframe
        allow="fullscreen; camera; microphone"
        class="css-15vps70"
        loading="lazy"
        referrerpolicy="no-referrer-when-downgrade"
        sandbox="allow-scripts allow-same-origin allow-downloads"
        src="undefined/solve/10/5"
      />
    </div>
  </div>
</div>
`;
